### PR TITLE
Data split dirichlet

### PR DESCRIPTION
**Pull request by 1henrypage**
Bug fixes also for author's code.

### CODE REVIEW

File: src/sampling.py [hunk 1]
```diff
@@ -22,6 +22,51 @@ def mnist_iid(dataset, num_users):
...
+    assert len(idxs_labels[1]) == 2 # sanity check
+
+    for label in uniq_labels:
+        relevant_idxs = idxs_labels[(idxs_labels[:, 1] == label)][:,0].T
+        proportions = np.random.dirichlet(np.full(num_users, beta))
+        splits = split_by_ratio(relevant_idxs, proportions)
+        for idx, split in enumerate(splits):
+            dict_users[idx] = np.concatenate([dict_users[idx], split])
+
+    return dict_users
```
Comment (tetra-code) [inline src/sampling.py:59]: Have you tested the method?

File: src/sampling.py [hunk 2]
```diff
@@ -22,6 +22,51 @@ def mnist_iid(dataset, num_users):
         all_idxs = list(set(all_idxs) - dict_users[i])
     return dict_users
 
+def split_dirichlet(dataset, num_users, is_cfar, beta=0.5):
```
Comment (tetra-code) [inline src/sampling.py]: It would be nice if you can add dynamic typing for each parameter and return values, especially when it comes to numpy arrays where dimensions can vary

File: src/sampling.py [hunk 3]
```diff
@@ -22,6 +22,51 @@ def mnist_iid(dataset, num_users):
...
+    their parameter name is different.
+    :return: dict mapping client id to idxs for training
+    """
+    idxs = np.arange(len(dataset))
+    labels = np.array(dataset.targets) if is_cfar else dataset.train_labels.numpy()
+    uniq_labels = np.unique(labels)
+    dict_users = {i: np.array([]) for i in range(num_users)}
+    idxs_labels = np.vstack((idxs, labels))
+    idxs_labels = idxs_labels[:, idxs_labels[1, :].argsort()]
+    idxs_labels = idxs_labels.transpose()
```
Comment (tetra-code) [inline src/sampling.py]: Why do need to transpose  the idxs_labels?

File: src/sampling.py [hunk 4]
```diff
@@ -22,6 +22,51 @@ def mnist_iid(dataset, num_users):
...
+
+    assert len(idxs_labels[1]) == 2 # sanity check
```
Comment (tetra-code) [inline src/sampling.py]: length of second column is labels correct? Why does this have to be size 2?

File: src/sampling.py [hunk 5]
```diff
@@ -22,6 +22,51 @@ def mnist_iid(dataset, num_users):
...
+
+    for label in uniq_labels:
+        relevant_idxs = idxs_labels[(idxs_labels[:, 1] == label)][:,0].T
+        proportions = np.random.dirichlet(np.full(num_users, beta))
```
Comment (tetra-code) [inline src/sampling.py:49]: Do we necessarily have to consider proportions for ALL users and not just some users?

File: src/utils.py [hunk 6]
```diff
@@ -57,9 +58,11 @@ def get_dataset(args):
                                       transform=apply_transform)
 
         # sample training data amongst users
-        if args.iid:
+        if args.iid == 1:
             # Sample IID user data from Mnist
             user_groups = mnist_iid(train_dataset, args.num_users)
+        elif args.iid == 2:
```
Comment (Lyadalachanchu) [inline src/utils.py:64]: Can you enable us to set the beta value as a parameter when calling it from the terminal (as an argument)?

File: src/sampling.py (hunk 3 above)
Comment (Lyadalachanchu) [reply src/sampling.py]: It's vertically stacked beforeand ig it looks nicer when its transposed (with idxs and labels as seperate columns?)

Comment (1henrypage) [reply src/sampling.py]: Exactly, it's used for the next section when filtering through the array. Once filtering is done, it is transposed back anyways.

File: src/sampling.py (hunk 4 above)
Comment (1henrypage) [reply src/sampling.py]: I changed the assertion to make it more clear what it asserts, it's not very important just an intermediary test to see that all the transposing when correcftl.y

File: src/sampling.py (hunk 5 above)
Comment (1henrypage) [reply src/sampling.py:49]: Yes. If this needs to be changed later it can be easily.

File: src/sampling.py (hunk 1 above)
Comment (1henrypage) [reply src/sampling.py:59]: Yes. Here's a split on 10 clients using MNIST. 

0: {0: 413, 1: 383, 2: 422, 3: 692, 4: 1229, 5: 2336, 6: 686, 7: 1073, 8: 169, 9: 21}, 
1: {0: 2757, 1: 683, 2: 331, 3: 377, 4: 1077, 5: 599, 6: 27, 7: 727, 8: 120}, 
2: {0: 496, 1: 4, 2: 1145, 3: 180, 4: 298, 5: 445, 6: 44, 7: 3, 8: 881, 9: 10}, 
3: {0: 635, 1: 2, 2: 90, 3: 164, 4: 521, 5: 74, 6: 599, 7: 84, 8: 213, 9: 1850}, 
4: {0: 28, 1: 841, 2: 269, 3: 454, 4: 4, 5: 563, 6: 25, 7: 667, 8: 1063, 9: 2376}, 
5: {0: 215, 2: 605, 3: 2402, 4: 1795, 5: 412, 6: 162, 7: 25, 8: 1076, 9: 461}, 
6: {0: 1, 1: 1819, 2: 227, 3: 1497, 4: 97, 5: 116, 6: 1166, 7: 614, 8: 1430, 9: 699}, 
7: {0: 51, 1: 579, 2: 2704, 3: 128, 4: 21, 5: 109, 6: 44, 7: 191, 8: 774, 9: 344}, 
8: {0: 126, 1: 2240, 2: 3, 3: 185, 4: 771, 5: 56, 6: 2030, 7: 62, 8: 63, 9: 65}, 
9: {0: 1201, 1: 191, 2: 162, 3: 52, 4: 28, 5: 711, 6: 1135, 7: 2819, 8: 61, 9: 123}}

File: src/federated_main.py [hunk 7]
```diff
@@ -29,8 +29,8 @@
     args = args_parser()
     exp_details(args)
 
-    if args.gpu_id:
-        torch.cuda.set_device(args.gpu_id)
+    if args.gpu:
```
Comment (Lyadalachanchu) [inline src/federated_main.py:32]: Change readme commands to use --gpu=cuda:0 instead of --gpu=0 (since it might trip others up)

File: src/utils.py (hunk 6 above)
Comment (1henrypage) [reply src/utils.py:64]: done

File: src/federated_main.py (hunk 7 above)
Comment (1henrypage) [reply src/federated_main.py:32]: Changed.

File: src/sampling.py [hunk 8]
```diff
@@ -23,6 +23,53 @@ def mnist_iid(dataset, num_users):
...
+            dict_users[idx] = np.concatenate([dict_users[idx], split])
+
+    return dict_users
+
+def split_by_ratio(arr, ratios):
+    """
+    Splits an np array according to some proportions, must sum to 1
+    """
+    arr = np.random.permutation(arr)
+    ind = np.add.accumulate(np.array(ratios) * len(arr)).astype(int)
```
Comment (Lyadalachanchu) [inline src/sampling.py:66]: Throw an error immediately if the size for any split is 0. Right now, if the size is zero, it doesn't throw the error until later (after some clients with data have already trained)

Comment (Lyadalachanchu) [reply src/sampling.py:66]: This happens whenever the training set has datapoints than the batch size. So not only when the split is 0, but even if it is relatively small (like 30 ish with the current 80,10,10 train test val split)

File: src/sampling.py (hunk 5 above)
Comment (tetra-code) [reply src/sampling.py:49]: Why is that?

Comment (1henrypage) [reply src/sampling.py:49]: Because it's easier for me?

File: src/sampling.py (hunk 8 above)
Comment (1henrypage) [reply src/sampling.py:66]: Are you talking about the split of 1 category or the split of users? I'm going to assume the second right? So I just enforce that each list in the dictionary must be longer than say 40?

Comment (Lyadalachanchu) [reply src/sampling.py:66]: Ah yes split of users mb. Should've put this comment for the method above

Comment (1henrypage) [reply src/sampling.py:66]: Ok I added this.

```python
for _ , dict_val in dict_users.items():
        if len(dict_val) < 40:
            # We just restart a split if a user isn't assigned enough samples.
            return split_dirichlet(dataset, num_users, is_cfar, beta)
```

Comment (Lyadalachanchu) [reply src/sampling.py:66]: looks good from my side then
